tests/xharness/Harness.cs

```diff
@@ -80,52 +80,104 @@ public class Harness : IHarness
+		public string WatchOSExtensionTemplate { get; private set; }
+		public string TodayContainerTemplate { get; private set; }
+		public string TodayExtensionTemplate { get; private set; }
+		public string BCLTodayExtensionTemplate { get; private set; }
+		public string MONO_PATH { get; private set; } // Use same name as in Makefiles, so that a grep finds it.
+		public string TVOS_MONO_PATH { get; private set; } // Use same name as in Makefiles, so that a grep finds it.
+		public bool INCLUDE_IOS { get; private set; }
+		public bool INCLUDE_TVOS { get; private set; }
+		public bool INCLUDE_WATCH { get; private set; }
+		public bool INCLUDE_MAC { get; private set; }
+		public string JENKINS_RESULTS_DIRECTORY { get; private set; } // Use same name as in Makefiles, so that a grep finds it.
+		public string MAC_DESTDIR { get; private set; }
+		public string IOS_DESTDIR { get; private set; }
+		public string MONO_IOS_SDK_DESTDIR { get; private set; }
+		public string MONO_MAC_SDK_DESTDIR { get; private set; }
+		public bool IncludeMac32 { get; private set; }
+		public bool ENABLE_XAMARIN { get; private set; }
 
 		// Run
-		public AppRunnerTarget Target { get; set; }
-		public string SdkRoot { get; set; }
-		public string Configuration { get; set; } = "Debug";
-		public string LogFile { get; set; }
-		public string LogDirectory { get; set; } = Environment.CurrentDirectory;
-		public double Timeout { get; set; } = 15; // in minutes
-		public double LaunchTimeout { get; set; } // in minutes
-		public bool DryRun { get; set; } // Most things don't support this. If you need it somewhere, implement it!
-		public string JenkinsConfiguration { get; set; }
-		public Dictionary<string, string> EnvironmentVariables { get; set; } = new Dictionary<string, string> ();
-		public string MarkdownSummaryPath { get; set; }
-		public string PeriodicCommand { get; set; }
-		public string PeriodicCommandArguments { get; set; }
-		public TimeSpan PeriodicCommandInterval { get; set; }
+		public AppRunnerTarget Target { get; }
+		public string SdkRoot { get; private set; }
+		public string Configuration { get; }
+		public string LogFile { get; }
+		public string LogDirectory { get; }
+		public double Timeout { get; } = 15; // in minutes
+		public double LaunchTimeout { get; } // in minutes
+		public bool DryRun { get; } // Most things don't support this. If you need it somewhere, implement it!
+		public string JenkinsConfiguration { get; }
+		public Dictionary<string, string> EnvironmentVariables { get; }
+		public string MarkdownSummaryPath { get; }
+		public string PeriodicCommand { get; }
+		public string PeriodicCommandArguments { get; }
+		public TimeSpan PeriodicCommandInterval { get; }
 		// whether tests that require access to system resources (system contacts, photo library, etc) should be executed or not
 		public bool? IncludeSystemPermissionTests { get; set; }
 
-		public Harness ()
-		{
+		public Harness (IProcessManager processManager,
+			HarnessAction action,
```

Comment:
A constructor with 26 required arguments is not a particularly pretty sight...

I think it would be much better to separate all the command-line argument data into a separate class, and just pass an instance of that class to the constructor.

Reply:
I am trying to solve following problem - someone initializing the class and forgetting to provide an argument.
If I enforce arguments in ctor, I can make them readonly and know that the state of my object is always correct.
I am not sure how else to enforce this other than method arguments.
If I make a wrapper argument class, this class will have to have this ctor and I haven't helped myself much?

I can do that though, if you think it's better.

Reply:
The idea of a wrapper class is something like this:

```csharp
class HarnessArguments {
    public bool AutoConf;
    public string Configuration;
    ...
}

class Harness {
    public bool AutoConf { get; }

    public Harness (HarnessArguments arguments)
    {
        AutoConf = arguments.AutoConf;

        // maybe verify that the arguments provided make sense.
    }
}
```

this way you can still make all the Harness fields readonly if you wish.

IMHO enforcing that all the arguments have been provided doesn't make much sense, because it's never the case that all arguments are needed, some are needed some times, and some are needed other times.

Reply:
I understand the idea.. I just don't like the fact that now if we add a new field, we must look for all places where we instantiate the arguments class and we won't know on compile time that we forgot to set a field somewhere. While this is perfectly solvable on compile time rather than runtime... I think here we really should aim for maximum correctness not visual appearance of the code and let the code tell us as early as possible something is wrong.

Anyway, I added the class..